flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -168,6 +168,33 @@ componentToExtendedValue(Fortran::lower::FirOpBuilder &builder,
   return component;
 }
 
+/// Given the address of an array element and the ExtendedValue describing the
+/// array, returns the ExtendedValue describing the array element. The purpose
+/// is to propagate the length parameters of the array to the element.
+static fir::ExtendedValue
+arrayElementToExtendedValue(Fortran::lower::FirOpBuilder &builder,
+                            mlir::Location loc, const fir::ExtendedValue &array,
+                            mlir::Value element) {
+  return array.match(
+      [&](const fir::CharBoxValue &cb) -> fir::ExtendedValue {
+        return cb.clone(element);
+      },
+      [&](const fir::CharArrayBoxValue &bv) -> fir::ExtendedValue {
+        return bv.cloneElement(element);
+      },
+      [&](const fir::BoxValue &bv) -> fir::ExtendedValue {
+        return bv.cloneElement(element);
+      },
+      [&](const fir::IrBoxValue &box) -> fir::ExtendedValue {
+        if (box.isCharacter())
+          return fir::CharBoxValue{element, builder.readCharLen(loc, box)};
+        if (box.isDerived())
+          TODO(loc, "get length parameters from IrBox");
```

Comment:
I think we can generate that code lazily. The distinction being

```fortran
subroutine s(arg, n, m)
  type(t(n,m)) :: arg
  ! [#]
```
and
```fortran
subroutine s(arg)
  class(*) :: arg
  ! [$]
  select type (a => arg)
    type is (t(*))
    !  [$$]
```

Here, at `[#]` we have some LEN type parameters that are specified in the callee. At `[$]` we have a box argument of polymorphic type, so we really have no idea if the box has LEN parameters. At `[$$]` though, we know the box's type and the the LEN parameters are present and specified in that boxed value.

This is just illustrative of tracking properties in lowering vs. properties that must come from the IR.

The orthogonal case is "mutability".
```fortran
  type t(p)
    integer, len :: p
    ...
  end type t
  type(t(n)),allocatable :: v
  ...
  allocate(t(m+4) :: v)
  ! [%]
```
Here `v` is a boxed value with a LEN parameter that can change at runtime. At `[%]` this LEN parameter is assigned the value of the expression `m+4`, of course.

Finally, we can have both the `[$]`, `[$$]` and `[%]` semantics simultaneously. For a given Fortran variable, _v_, we may both need to defer (read lazily) its properties (such as LEN parameters, bounds, etc.) _and_ we may need to dynamically set them to runtime computed values.

Reply:
I am generally in line with the lazy approach, I think we are doing it when mapping symbols to MutableBoxValue/IrBoxValue (we do not read the fir.box at that point anymore).

However, I am not sure I get how and why you want to delay the fir.box reading in this very specific piece of code:

This routine is building an Exv for an array element reference, so that mean we are in the middle of an array element refence, which implies:
- the length cannot be changed of the middle of the statement it belongs to.
- the result of this will never be a mutableBox (one cannot set the properties of an array element individually).
- the result is a scalar, so unless the array is polymorphic, we do not need a fir.box to describe/work with it.
- the fir.box we have cannot be kept to describe the result, it describes the array, not the element.
- at that point, the array cannot be unlimited polymorphic/assumed rank anymore (otherwise the array element ref would be illegal semantically).

So my approach here is to read greedily the lengths to create a BoxValue for the element Exv, and let dead code elimination get rid of the unused one (e.g. if the Exv consumer does generate any code using them).

Here is the typical use case for it:
```
subroutine foo(x)
  type(t(*)) :: x(:, :)
  call bar(x(1, 5))
end subroutine
```

- `x` is lowered to an IrBoxValue, we do not read its length parameter on entry (so we are lazy here).
- `arrayElementToExtendedValue` will be called after computing the address of `x(1, 5)`, it will build a BoxValue with the base address and the length parameter of `x(1, 5)`. It reads the length parameter value from the fir.box in the array IrBoxValue of `x`.
- We would then embox this BoxValue in a scalar fir.box to pass that bar in the call.